crates/bevy_ecs/src/schedule/executor_parallel.rs

```diff
@@ -167,7 +167,7 @@ impl ParallelExecutor {
     fn prepare_systems<'scope>(
         &mut self,
         scope: &mut Scope<'scope, ()>,
-        systems: &'scope mut [ParallelSystemContainer],
+        systems: &'scope mut [SystemContainer],
         world: &'scope World,
```

Comment:
Just thinking out loud here. For stageless, I'm pretty sure this will need to become `&'scope UnsafeCell<World>` so the executor can support exclusive systems (can't just hold the initial `&mut World` reference forever).

AFAIK there is no "read-only" schedule type, so it should be fine. But is there anything like that in the rendering crates? If so, it would require a `MaybeUnsafeCellRef` enum/union like 4166 has. If not, just something to keep in mind.

Reply:
> Just thinking out loud here. For stageless, I'm pretty sure this will need to become &'scope UnsafeCell<World> so the executor can support exclusive systems (can't just hold the initial &mut World reference forever).

Why can't the `Executor::run` take a `&mut World` reference, then convert to `&World` when executing groups of parallel systems? We'll need a provable `&mut World` sync point anyway, why not just encode that directly in the logic / type system?

Reply:
> Why can't the Executor::run take a `&mut World` reference, then convert to `&World` when executing groups of parallel systems?

You misunderstood me. What you wrote is exactly what I expect. `Executor::run` *has* to take `&mut World`. I'm just pointing out that we can't hold that `&mut World` reference the whole time. We have to downgrade it into `&UnsafeCell<World>`, so we can make `&World` and `&mut World` as needed.

I know `bevy_render` employes read-only logic like [`SystemState::get`][1] a lot, so I'm just saying that you would not be able to reuse this `Executor::run` for some schedule-level equivalent of that. You would have to duplicate the impl or introduce a union/enum type, because `&World` -> `&UnsafeCell<World>` is not legal (yet).

[1]: https://docs.rs/bevy/latest/bevy/ecs/system/struct.SystemState.html#method.get

Reply:
> We have to downgrade it into `&UnsafeCell<World>`.

Like `Executor::run` needs to behave more or less like this for stageless.

```rust
fn run(&mut self, schedule: &mut Schedule, world: &mut World) {
    ComputeTaskPool::init(TaskPool::default).scope(|scope| {
        let mut executor = async {
            // systems with zero dependencies are ready
            for (sys_idx, sys_meta) in self.system_task_metadata.iter_mut().enumerate() {
                if sys_meta.dependencies_total == 0 {
                    self.ready_systems.insert(sys_idx);
                }
            }

            // We can't hold onto the `&mut World` reference inside this loop.
            // We have to downgrade it into `&UnsafeCell<World>`.
      
            // SAFETY: We are allowed to turn `&mut T` into `&UnsafeCell<T>`.
            let world = unsafe { &*(world as *mut World as *const UnsafeCell<World>) };
            
            while self.completed_systems.len() != schedule.systems.len() {
                // spawn tasks for ready systems (exclusive systems can just be run "inline")
                self.spawn_system_tasks(scope, schedule, world).await;

                if self.running_systems.count_ones(..) != 0 {
                    // wait for at least one system to complete
                    let index = self.finish_receiver.recv().await.unwrap();
                    self.finish_system_and_signal_dependents(index);
                    while let Ok(index) = self.finish_receiver.try_recv() {
                        self.finish_system_and_signal_dependents(index);
                    }

                    // remove the access of the completed systems
                    self.active_access.clear();
                    for sys_idx in self.running_systems.iter() {
                        let sys_meta = &self.system_task_metadata[sys_idx];
                        self.active_access
                            .extend(&sys_meta.archetype_component_access);
                    }
                }

                if !self.exclusive_running && self.active_access.is_empty() {
                    // SAFETY: no active borrow on the `World` itself or data inside it
                    let world = unsafe { &mut *world.get() };
                    self.check_apply_system_buffers(schedule, world);
                }
            }
            self.completed_systems.clear();
        };
        scope.spawn(runner);
    });
}
```

Reply:
To be clear, I'm not asking you to change anything. I was just checking if going with this instead of 4166 would affect other parts of the stageless implementation work.

Reply:
> You misunderstood me. What you wrote is exactly what I expect. Executor::run has to take &mut World. I'm just pointing out that we can't hold that &mut World reference the whole time. We have to convert it into a &UnsafeCell<World> (edit: or *mut World), so we can make &World and &mut World as needed.

In that case I don't think I misunderstood you. To clarify my second sentence further: I don't think we should ever need to convert `&mut World` to an UnsafeCell. We will only ever be running exclusive systems (with `&mut World`) or groups of parallel systems (with `&World`). In that context, I see no reason why we can't express that with normal rust borrows, stemming from the root `&mut World` passed in to the executor.

I don't see any reason why your example needs the UnsafeCell. Forgive the pseudocode, but this pattern does compile just fine:

```rust
fn run(&mut self, schedule: &mut Schedule, world: &mut World) {
    ComputeTaskPool::init(TaskPool::default).scope(|scope| {
        let mut executor = async {
            while !self.is_finished() {
                spawn_system_tasks(scope, world).await;
            }
        };
        // scope.spawn(executor);
    });
}

async fn spawn_system_tasks(&mut self, scope: &mut Scope<'_, ()>, world: &mut World) {
    if self.is_time_to_run_exclusive() {
        self.run_ready_exclusive(scope, world).await
    }

    if self.is_time_to_run_parallel() {
        self.run_ready_parallel(world).await
    }
}
async fn run_ready_parallel(&mut self, scope: &mut Scope<'_, ()>, world: &mut World) {}
async fn run_ready_exclusive(&mut self, world: &World) {}
```

Reply:
Oops I flipped the borrows in the `parallel` and `exclusive` functions. Editing now to fix that :)

Reply:
We can spawn tasks for parallel systems while other parallel systems are running, so isn't it UB to still have an active `&mut World` reference? Like, how are you able to call `spawn_system_tasks` in your example while other systems are running.

Reply:
Like having to wait for all running systems to complete before we can start new systems seems like an unnecessary loss in performance just avoid using an `UnsafeCell` here.

Reply:
Ok I see the problem you're outlining. As written, the example above wouldn't solve that problem. But this would:

```rust
fn run(&mut self, schedule: &mut Schedule, world: &mut World) {
    ComputeTaskPool::init(TaskPool::default).scope(|scope| {
        let mut executor = async {
            while !self.is_finished() {
                 self.run_ready_parallel_tasks_until_a_sync_point_is_reached(scope, world).await;
                 self.run_ready_exclusive_systems_until_a_sync_point_is_reached(world).await;
            }
        };
    });
}
```

Reply:
Hmm, I'm not sure it's that simple. The executor needs to evaluate conditions as well. TBD what ends up being the most readable solution. There's no difference between 4166/6083 on this particular point though, so you can mark this thread as resolved.

Reply:
Yeah I'm on board for the targeted UnsafeCell usage in the executor if the alternative is too unwieldy, but I think its worth trying to lean on rust here if we can / it isn't significantly worse.

Happy to table this convo until later!